### PR TITLE
`Assessment`: Fix an issue with complaints

### PR DESCRIPTION
**Pull request by akesfeden**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Addresses the issue nr. 6 at [WS22/23 - Issues Assessment Subteam](https://confluence.ase.in.tum.de/pages/viewpage.action?pageId=148579053). Bug descirption copied from there:

> When you assess a complaint and add a new feedback box but do not enter feedback text but only points and click submit, you get an alert that information is missing but the complaint is submitted and you cannot change it anymore.
> 
> Context I was an instructor.

### Description

`ComplaintsForTutorComponent`, delegates the server call to the parent endpoint when an assessment is saved after a complaint, However, this component assumes the parent component always completes the saving operation successfully and marks the complaint as handled, lock as released and disables the accept/reject buttons (only on UI, no additional server call is made). This PR includes `onSuccess`/onError` functions in the output for the parent components to allow `ComplaintsForTutorComponent` receive feedback for the assessment validation and save operations and adapt accordingly.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Submitted Exam with  at least 1 text, 1 modeling, 1 file upload and 1 programming exercise. (Exercises can be in different exams)

1. Log in to Artemis
2. Navigate to Course Administration
3. As an Instructor, evaluate 1 text, 1 modeling, 1 file upload and 1 programming exercise.
4. As a Student, make a complaint to each one of the above exercises.
5. As an Instructor, add an invalid feedback (e.g. for text exercise, you can add a new feedback box but do not enter feedback text but only points) and accept the complaint.
6. Make sure that there is an error alert shown but the accept/reject buttons are still enabled and the UI still shows the lock is held.
7. Fix the invalid feedback and accept the complaint, ensure that the operation works successfully as expected.

#### Exam Mode Testing
Changes only involve assessment mode and covered by the testing steps above.

#### Code Review
- [x] Review 1
#### Manual Tests
- [x] Test 1
- [x] Test 2
- [x] Test 3

#### From Codacy

| [Impacted Files](https://codecov.io/gh/ls1intum/Artemis/pull/6202?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum) | Coverage Δ | |
|---|---|---|
| [complaints-for-tutor.component.ts](https://codecov.io/gh/ls1intum/Artemis/pull/6202?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum#diff-c3JjL21haW4vd2ViYXBwL2FwcC9jb21wbGFpbnRzL2NvbXBsYWludHMtZm9yLXR1dG9yL2NvbXBsYWludHMtZm9yLXR1dG9yLmNvbXBvbmVudC50cw==) | `77.57% <100.00%> (+0.64%)` | :arrow_up: |
| [file-upload-assessment.component.ts](https://codecov.io/gh/ls1intum/Artemis/pull/6202?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum#diff-c3JjL21haW4vd2ViYXBwL2FwcC9leGVyY2lzZXMvZmlsZS11cGxvYWQvYXNzZXNzL2ZpbGUtdXBsb2FkLWFzc2Vzc21lbnQuY29tcG9uZW50LnRz) | `83.96% <100.00%> (+1.06%)` | :arrow_up: |
| [modeling-assessment-editor.component.ts](https://codecov.io/gh/ls1intum/Artemis/pull/6202?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum#diff-c3JjL21haW4vd2ViYXBwL2FwcC9leGVyY2lzZXMvbW9kZWxpbmcvYXNzZXNzL21vZGVsaW5nLWFzc2Vzc21lbnQtZWRpdG9yL21vZGVsaW5nLWFzc2Vzc21lbnQtZWRpdG9yLmNvbXBvbmVudC50cw==) | `67.44% <100.00%> (+1.81%)` | :arrow_up: |
| [code-editor-tutor-assessment-container.component.ts](https://codecov.io/gh/ls1intum/Artemis/pull/6202?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum#diff-c3JjL21haW4vd2ViYXBwL2FwcC9leGVyY2lzZXMvcHJvZ3JhbW1pbmcvYXNzZXNzL2NvZGUtZWRpdG9yLXR1dG9yLWFzc2Vzc21lbnQtY29udGFpbmVyLmNvbXBvbmVudC50cw==) | `75.37% <100.00%> (+2.47%)` | :arrow_up: |
| [text-submission-assessment.component.ts](https://codecov.io/gh/ls1intum/Artemis/pull/6202?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum#diff-c3JjL21haW4vd2ViYXBwL2FwcC9leGVyY2lzZXMvdGV4dC9hc3Nlc3MvdGV4dC1zdWJtaXNzaW9uLWFzc2Vzc21lbnQuY29tcG9uZW50LnRz) | `82.32% <100.00%> (+2.32%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/ls1intum/Artemis/pull/6202?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`

#### Screenshot of the error
image

### CODE REVIEW

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Code looks good apart from this:

File: src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts [hunk 1]
```diff
@@ -351,21 +351,32 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
...
 
         this.assessmentsService
             // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
-            .updateAssessmentAfterComplaint(this.assessments, this.textBlocksWithFeedback, complaintResponse, this.submission?.id!, this.participation?.id!)
+            .updateAssessmentAfterComplaint(
+                this.assessments,
+                this.textBlocksWithFeedback,
+                assessmentAfterComplaint.complaintResponse,
+                this.submission?.id!,
+                this.participation?.id!,
```
[the remark this answers is not in the document]
Comment (akesfeden) [reply src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts]: Thanks, I moved eslint-disable comment to correct lines.

**Review by Vinicius-Agreste — CHANGES_REQUESTED** [review]
Tested on ts1.Programming and modelling exercises did not work as expected (Text and File Upload worked flawlessly). I was able to accept (solve) the complaint in both while having an incomplete feedback.

**Comment by akesfeden** [comment]
@Vinicius-Agreste can you share the urls of the unexpected programming and modeling assessments on TS1?

**Comment by akesfeden** [comment]
@Vinicius-Agreste I think I got the issue. This PR was about preventing the UI to be in an unusable state with disabled when there a complaint is accepted with invalid feedback, but the issue you found seems to be about determining when does a feedback considered invalid and raising an error according to that (more related to the feedbacks themselves rather than the complaint). But I can try to address that here too.

**Comment by akesfeden** [comment]
@Vinicius-Agreste The validations for programming and modeling exercises are added in https://github.com/ls1intum/Artemis/pull/6202/commits/ef97bc5c9b4a294f29d94c51f08c78f55fa35543, could you check again?

**Review by laadvo — APPROVED** [review]
Tested using an exam with all 4 different types of exercises, everything is working as expected. Tested on ts1.

**Review by Vinicius-Agreste — COMMENTED** [review]
@akesfeden thank you for addressing my review. I tested it again and got the same result.
 If I understood it correctly, the apparent bug I faced is not really related to your PR. If you'd like, I can approve your PR and explain the bug I found in greater detail somewhere else. Let me know.
I went ahead and tested on a new Exam and noticed that when accepting a complaint with an incomplete feedback (without text) the feedback's points are added to the exercise calculation. See the following example: In this modelling exercise I added the first feedback "no" adding 0.5 points. Then, the student submits a complaint. I create a new, incomplete feedback (-0.5 points) and accept the complaint. The complaint gets resolved, and the -0.5 points are added to the exercise result (you can see the empty feedback message in the exercise result student view).
image

image

**Review by Vinicius-Agreste — APPROVED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Thanks for addressing my comment. Code looks good

**Pull request merged**